### PR TITLE
CRD versioning design proposal [No Conversion]

### PR DESCRIPTION
**Pull request by mbohlool**
CRD versioning proposal is discussed in detail here:

https://docs.google.com/document/d/1Ucf7JwyHpy7QlgHIN2Rst_q6yT0eeN9euzUV6kte6aY/edit

This is the main no conversion part of the design agreed on by the community/API machinery SIG.

@sttts @deads2k @lavalamp

### CODE REVIEW

**Comment by deads2k** [comment]
I think this matches what we talked about before.

@liggitt chance for final comments.  I'll probably tag on Monday.

@mbohlool there may be a little bikeshedding on Name versus Version in the substruct, but in looks well scoped.

**Comment by deads2k** [comment]
> @liggitt chance for final comments. I'll probably tag on Monday.

@sttts you too.

**Comment by mbohlool** [comment]
> there may be a little bikeshedding on Name versus Version in the substruct, but in looks well scoped.

'Version.Version' vs 'Version.Name'? I vote for `Version.Name` :)

**Comment by liggitt** [comment]
> @liggitt chance for final comments. I'll probably tag on Monday.

no other comments from me

File: contributors/design-proposals/api-machinery/customresources-versioning.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+```
+
+# **Validation**
+
+Basic validations needed for the `version` field are:
+
+* `Spec.Version` field exists in `Spec.Versions` field.
+* The version defined in `Spec.Version` field should point to a `Served` Version in `Spec.Versions` list except when we do not serve any version (i.e. all versions in `Spec.Versions` field are disabled by `Served` set to `False`). This is for backward compatibility. An old controller expect that version to be served but only the whole CRD is served. CRD Registration controller should unregister a CRD with no serving version.
+* None of the `Status.StoredVersion` can be removed from `Spec.Versions` list.
+* Only one of the versions in `spec.Versions` can flag as `Storage` version.
```
Comment (sttts) [inline contributors/design-proposals/api-machinery/customresources-versioning.md]: Exactly one...

**Comment by sttts** [comment]
Lgtm.

**Pull request merged**